pkg/storage/replica.go

```diff
@@ -6019,26 +6019,20 @@ func calcReplicaMetrics(
 	m.Leader = isRaftLeader(raftStatus)
 	m.Quiescent = quiescent
 
-	// We gather per-range stats on either the leader or, if there is no leader,
-	// the first live replica in the descriptor. Note that the first live replica
-	// is an arbitrary choice. We want to select one live replica to do the
-	// counting that all replicas can agree on.
+	// We gather per-range stats from the first live replica in the descriptor.
+	// Note that the first live replica is an arbitrary choice. We want to select
+	// one live replica to do the counting that all replicas can agree on.
 	//
 	// Note that the current heuristics can double count. If the first live
 	// replica is on a node that is partitioned from the other replicas in the
-	// range it may not know the leader even though there is one. This scenario
-	// seems rare as it requires the partitioned node to be alive enough to be
-	// performing liveness heartbeats.
-	if !HasRaftLeader(raftStatus) {
-		// The range doesn't have a leader or we don't know who the leader is.
-		for _, rd := range desc.Replicas {
-			if livenessMap[rd.NodeID] {
-				m.RangeCounter = rd.StoreID == storeID
-				break
-			}
+	// range, there may be multiple nodes which believe they are the first live
+	// replica. This scenario seems rare as it requires the partitioned node to be
+	// alive enough to be performing liveness heartbeats.
+	for _, rd := range desc.Replicas {
+		if livenessMap[rd.NodeID] {
+			m.RangeCounter = rd.StoreID == storeID
```

Comment:
`RangeCounter` is used for more than just the range count metric. Look a couple lines below - this will have the side effect of making metrics like `Unavailable` and `Underreplicated` be calculated on the non-raft leader even when there is a raft leader, which I don't think we want.

Am I misunderstanding something? If not, we should be able to restructure this to still use the raft leader in possible for the metrics that depend on `raftStatus` while using the "smallest store ID" logic for the range count.

Reply:
Non-leader ranges can still estimate under-replicated and unavailable ranges using the liveness table; on the Leader range, the estimate is slightly enhanced by checking the log progress of each replica.  That improved estimation is what we lose

However, as we see from the original issue for this change, it's very likely that a large number of ranges in the cluster will be in a state where the old algorithm (leader or smallest ID) will not count anything for a large number of ranges.

So this is a tradeoff between:

* Losing the improved estimation for "behindRanges" in more circumstances
* Not counting under-replicated or unavailable ranges which are in a partially leaderless state.

Reply:
As I said offline:

I should have looked more closely at the code originally . I saw that `calcGoodReplicas` relied on the replica’s `raftStatus` and assumed that it completely relied on it rather than falling back to using node liveness. That would have been bad since only the leader has knows about all the replicas’ progress states. Given that we fall back to liveness, I’m ok with this tradeoff.

Reply:
Uggh, Github cut off the rest of my comment somehow. I wanted to also say:

However, given this tradeoff, I still think there's some code shuffling that needs to be done to:

* Never use the `raftStatus` when calculating the number of live replicas (so that results don't change if the leader happens to be or not be the first live replica in the descriptor)
* Always calculate `BehindCount` on the leader regardless of whether it's the first live replica in the descriptor (so that `BehindCount` doesn't depend on which replica is the leader)